[xbmc] Turn DllLibCurl into a proper curl wrapper.

Paxxi opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

This is part 2 of #13360 and aims to turn DllLibCurl into a proper wrapper class for Curl.
Most of the work is done in the first commit before moving files to make the review easier.

### Design Goals ###
Wrap Curl to avoid curl types and curl headers leaking into the rest of the project.
Add an abstraction over Curl that's hopefully easier to use.

### Details ###
This renames DllLibCurl to CCurl and moves it into platform instead of filesystem. I was thinking about adding an externals folder but with curl maybe ending up being the only wrapper it made little sense.

g_curlInterface and it's caching of connections is gone to avoid global state management that at best provides a negligible performance increase.

A very simple class `CCurlGlobalInit` is added that's only job is to init/deinit curl globally and set up locks for static linking.

### Doubt ###
After I started on this I started doubting myself a bit so curious about others thoughts on this. Same goal can probably be achieved with a somewhat redesigned `CCurlFile`.

Maybe use `std::error_condition` instead of `std::error_code` as I learned that technically std::error_code is for system errors and std::error_condition is for "other" errors.

## Description
<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Rechi commented on xbmc/platform/Curl.cpp:
```diff
@@ -0,0 +1,663 @@
+#if LIBCURL_VERSION_NUM >= 0x072000 // 0.7.32
+  ec = make_error_code(curl_easy_setopt(m_session->m_easy, CURLOPT_XFERINFOFUNCTION, cb));
+#else
+  ec = make_error_code(curl_easy_setopt(m_session->m_easy, CURLOPT_PROGRESSFUNCTION, cb);
```
`..., cb));`

Rechi commented on xbmc/platform/Curl.cpp:
```diff
@@ -0,0 +1,663 @@
+  if (!referer.empty())
+    return make_error_code(curl_easy_setopt(m_session->m_easy, CURLOPT_REFERER, referer.c_str()));
+
+  auto ec = make_error_code(curl_easy_setopt(m_session->m_easy, CURLOPT_REFERER, nullptr));
```
`..., CURLOPT_REFERER, NULL));` to get it compiled for windows uwp

Paxxi replied:
Interesting. Didn't try building for uwp, so it handles C varargs differently?

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -192,12 +172,12 @@ size_t CCurlFile::CReadState::ReadCallback(char *buffer, size_t size, size_t nit
 size_t CCurlFile::CReadState::WriteCallback(char *buffer, size_t size, size_t nitems)
 {
-  unsigned int amount = size * nitems;
+  auto amount = size * nitems;
```
auto for a simple type like `size_t`?

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -1125,26 +1009,26 @@ ssize_t CCurlFile::Write(const void* lpBuf, size_t uiBufSize)
     {
-      long code;
-      if(g_curlInterface.easy_getinfo(m_state->m_easyHandle, CURLINFO_RESPONSE_CODE, &code) == CURLE_OK )
+      auto code = m_state->m_curl.GetResponseCode(ec);
```
auto for long?

Rechi reviewed (commented):
I don't like the usage of auto for fundamental types (+ `size_t`)

ccope commented:
re: connection caching, have you done a comparison with a large library import? I could see caching help improve throughput when fetching a lot of images

lrusak commented:
@Paxxi do you have any intention on working on this some more? It would be nice to get this in for V19 :smile_cat:

Paxxi commented:
I abandoned it because I wasn't sure about the design. If you believe it's useful I can probably get it into shape 😀

lrusak commented:
> I abandoned it because I wasn't sure about the design. If you believe it's useful I can probably get it into shape

I don't have any issues with the design :+1: If it works the same then it's nice to have the cleanup done.

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -41,53 +42,28 @@ using namespace XCURL;
-};
+CurlProxy ProxyType2CurlProxyType(CCurlFile::ProxyType proxy) noexcept
+{
+  if (!g_advancedSettings.CanLogComponent(LOGCURL))
```
either the conditional statement for the if is missing or the below switch is indented incorrectly

Paxxi replied:
yeah I removed it in next commit, was a merge error

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -454,202 +410,137 @@ void CCurlFile::Close()
-  g_curlInterface.easy_setopt(h, CURLOPT_LOW_SPEED_TIME, m_lowspeedtime);
+  curl.SetLowSpeedLimit(1, m_lowspeedtime);
 
+    g_curlInterface.easy_setopt(h, CURLOPT_IGNORE_CONTENT_LENGTH, 1);
```
wrong indentation

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -1135,26 +1014,26 @@ ssize_t CCurlFile::Write(const void* lpBuf, size_t uiBufSize)
-      long code;
-      if(g_curlInterface.easy_getinfo(m_state->m_easyHandle, CURLINFO_RESPONSE_CODE, &code) == CURLE_OK )
+      auto code = m_state->m_curl.GetResponseCode(ec);
+      if(!ec)
```
space is missing after `if`

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -1212,42 +1091,38 @@ bool CCurlFile::Exists(const CURL& url)
-    long code;
-    if(g_curlInterface.easy_getinfo(m_state->m_easyHandle, CURLINFO_RESPONSE_CODE, &code) == CURLE_OK && code != 404 )
+    long code = m_state->m_curl.GetResponseCode(ec);
+    if(!ec && code != 404 )
```
space is missing after `if`

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -1433,27 +1300,27 @@ int CCurlFile::Stat(const CURL& url, struct __stat64* buffer)
+  auto result = m_state->m_curl.EasyPerform();
 
-  if(result == CURLE_HTTP_RETURNED_ERROR)
+  if(result == CurlCode::HTTP_RETURNED_ERROR)
```
space is missing after `if`

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -1433,27 +1300,27 @@ int CCurlFile::Stat(const CURL& url, struct __stat64* buffer)
-    long code;
-    if(g_curlInterface.easy_getinfo(m_state->m_easyHandle, CURLINFO_RESPONSE_CODE, &code) == CURLE_OK && code == 404 )
+    auto code = m_state->m_curl.GetResponseCode(ec);
+    if(!ec && code == 404 )
```
space is missing after `if`

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -1473,64 +1340,52 @@ int CCurlFile::Stat(const CURL& url, struct __stat64* buffer)
   }
 
-  if( result != CURLE_ABORTED_BY_CALLBACK && result != CURLE_OK )
+  if(result && result != CurlCode::ABORTED_BY_CALLBACK)
```
space is missing after `if`

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -1473,64 +1340,52 @@ int CCurlFile::Stat(const CURL& url, struct __stat64* buffer)
+
+    memset(buffer, 0, sizeof(struct __stat64));
+    buffer->st_size = length;
+    if(!content.empty() && content.compare("text/html") == 0) //consider html files directories
```
space is missing after `if`

Rechi commented on xbmc/filesystem/CurlFile.h:
```diff
@@ -11,29 +11,24 @@
 #include "utils/HttpHeader.h"
 #include "utils/RingBuffer.h"
-
+#include "DllLibCurl.h"
```
wrong include order

Rechi commented on xbmc/filesystem/DllLibCurl.cpp:
```diff
@@ -8,307 +8,668 @@
+#include <system_error> // for error_code, error_category
+#include <type_traits>  // for move
+#include <utility>      // for pair
+#include <vector>       // for vector
```
> `Do not use whitespace to vertically align`

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#35-vertical-alignment

Rechi commented on xbmc/filesystem/DllLibCurl.h:
```diff
@@ -8,93 +8,304 @@
-#include <vector>
+#include <cstddef>      // for size_t
+#include <cstdint>      // for int64_t
+#include <string>       // for string
```
wrong include order

Rechi commented on xbmc/filesystem/DllLibCurl.h:
```diff
@@ -8,93 +8,304 @@
+{
+  DEFAULT,   /* let libcurl pick */
+  MULTICWD,  /* single CWD operation for each path part */
+  NOCWD,     /* no CWD at all */
```
> `Do not use whitespace to vertically align`

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#35-vertical-alignment

Rechi commented on xbmc/filesystem/DllLibCurl.h:
```diff
@@ -8,93 +8,304 @@
+{
+  DEFAULT, /* let libcurl decide */
+  SSL,     /* use "AUTH SSL" */
+  TLS,     /* use "AUTH TLS" */
```
> `Do not use whitespace to vertically align`

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#35-vertical-alignment

Rechi commented on xbmc/filesystem/DllLibCurl.h:
```diff
@@ -8,93 +8,304 @@
+  DIGEST_IE = 16,
+  NTLM_WB = 32,
+  ONLY = 1u << 31,
+  ANY = 1|2|4|8|32,
```
spaces are missing around `|` operators

Rechi commented on xbmc/filesystem/DllLibCurl.h:
```diff
@@ -8,93 +8,304 @@
+  NTLM_WB = 32,
+  ONLY = 1u << 31,
+  ANY = 1|2|4|8|32,
+  ANYSAFE = 2|4|8|32
```
spaces are missing around `|` operators

Rechi commented on xbmc/filesystem/DllLibCurl.h:
```diff
@@ -8,93 +8,304 @@
+  std::error_code MultiFdSet(fd_set* read_fd_set,
+                              fd_set* write_fd_set,
+                              fd_set* exc_fd_set,
+                              int* max_fd);
```
indentation is once space off

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -31,33 +31,35 @@
 #endif
 
-#include "DllLibCurl.h"
+#include "platform/Curl.h"
```
wrong include order

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -247,12 +249,12 @@ bool CCurlFile::CReadState::Seek(int64_t pos)
     if (FillBuffer(m_bufferSize) != FILLBUFFER_OK)
     {
-      if(!m_buffer.SkipBytes(-len))
+      if(!m_buffer.SkipBytes(-static_cast<int>(len)))
```
space is missing after `if`

Rechi commented on xbmc/filesystem/CurlFile.cpp:
```diff
@@ -262,7 +264,7 @@ bool CCurlFile::CReadState::Seek(int64_t pos)
     {
       CLog::Log(LOGERROR, "%s - Failed to skip to position after having filled buffer", __FUNCTION__);
-      if(!m_buffer.SkipBytes(-len))
+      if(!m_buffer.SkipBytes(-static_cast<int>(len)))
```
space is missing after `if`

Rechi commented on xbmc/filesystem/CurlFile.h:
```diff
@@ -11,7 +11,7 @@
 #include "utils/HttpHeader.h"
 #include "utils/RingBuffer.h"
-#include "DllLibCurl.h"
+#include "platform/Curl.h"
```
wrong include order

Rechi commented on xbmc/filesystem/DAVFile.cpp:
```diff
@@ -9,14 +9,14 @@
 #include "utils/log.h"
+#include "platform/Curl.h"
+#include "utils/XBMCTinyXML.h"
+#include "utils/RegExp.h"
```
wrong include order

Rechi commented on xbmc/platform/Curl.h:
```diff
@@ -12,40 +12,16 @@
+#include <cstdint>       // for int64_t
+#include <cstddef>       // for size_t
+#include <sys/select.h>  // for fd_set
+#include <string>        // for string
```
wrong include order

Rechi commented on xbmc/platform/Curl.h:
```diff
@@ -192,16 +168,16 @@ enum CurlFtpAuth
-  ANY = 1|2|4|8|32,
-  ANYSAFE = 2|4|8|32
+  ANY = ~16u,
+  ANYSAFE = ~(1u|16u)
```
spaces around `|` operator are missing

Rechi commented on xbmc/platform/Curl.h:
```diff
@@ -297,15 +277,28 @@ class CCurl
+  std::error_code AddHeader(const std::string& header);
 
 private:
+
```
superfluous empty line

Rechi commented on xbmc/platform/xbmc.cpp:
```diff
@@ -23,11 +23,16 @@
 #endif
 
 #include "platform/MessagePrinter.h"
+#include "platform/CurlGlobalInit.h"
```
wrong include order

Rechi commented on xbmc/filesystem/CurlFile.h:
```diff
@@ -9,38 +9,41 @@
+    void Close() override;
+    bool ReadString(char* szLine, int iLineLength) override
+    {
+      return m_state->ReadString(szLine, iLineLength); }
```
move curly bracket into new line and fix indentation of the lines below

Paxxi commented:
@lrusak rebased and quickly tested, installed some add-ons. Needs more testing for various protocols and platforms

lrusak commented:
@Paxxi how can I help push this through?

Paxxi commented:
@lrusak I guess testing it would be the main thing to be somewhat certain it doesn't horribly break things 😄

lrusak commented:
> @lrusak I guess testing it would be the main thing to be somewhat certain it doesn't horribly break things

@Paxxi isn't that what tests are for? :joy: 
Maybe we can get @MilhouseVH including this?

Rechi reviewed (commented):
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

MilhouseVH commented:
Multiple reports of Netflix and other video add-ons no longer working with this PR: [forum](https://forum.kodi.tv/showthread.php?tid=343069&pid=2924701#pid2924701)

Without this PR, the Netflix addon is working normally.

Paxxi commented:
Thanks @MilhouseVH 😀
Need to look at that then

phunkyfish commented:
@Paxxi will you be continuing with this PR or would you like to out it on the backburner (i.e. labelled as such and closed).

Paxxi commented:
Plan is to get it merged. Need to figure out why some add-ons aren't working and it should be good to go

Paxxi committed:
Move DllLibCurl to platform

Paxxi committed:
clang-format to minimize diff

Paxxi committed:
Change DllLibCurl into a wrapper class for curl.

This encapsulates all of curl and tries to avoid bits leaking out
into the rest of the project.
